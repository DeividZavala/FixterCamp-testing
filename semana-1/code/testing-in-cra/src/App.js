import React, { useState, useRef } from "react";
import "./App.css";

function App() {
  const [elements, setElements] = useState([]);
  const [error, setError] = useState();
  const inputRef = useRef();
  const addElement = () => {
    const { value } = inputRef.current;
    if (!value) return setError("Debes agregar texto");
    setElements((prev) => [...prev, value]);
    inputRef.current.value = "";
    setError(undefined);
  };
  return (
    <div className="App">
      {elements.length ? (
        <ul>
          {elements.map((element, i) => (
            <li key={i}>{element}</li>
          ))}
        </ul>
      ) : (
        <h1>Sin elementos para mostrar</h1>
      )}
      <div>
        <input ref={inputRef} placeholder="Ingresa el texto" />
        <button onClick={addElement}>agregar</button>
        {error && <p>{error}</p>}
      </div>
    </div>
  );
}

export default App;
