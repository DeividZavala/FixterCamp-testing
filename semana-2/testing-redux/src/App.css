.App {
  text-align: center;
}

.cart {
  position: relative;
}

.cart .uk-badge {
  position: absolute;
  top: -35%;
  left: 60%;
  z-index: 999;
}
