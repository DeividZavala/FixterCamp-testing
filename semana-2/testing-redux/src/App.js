import React from "react";
import Header from "./components/Header";
import Card from "./components/Card";
import "./App.css";

function App() {
  return (
    <div className="App">
      <Header />
      <div className="uk-section">
        <div className="uk-container uk-container-expand">
          <div className="uk-grid uk-grid-small">
            <div className="uk-width-2-3">
              <div className="uk-grid uk-child-width-1-3 uk-grid-small">
                <Card />
                <Card />
                <Card />
                <Card />
                <Card />
                <Card />
              </div>
            </div>
            <div className="uk-width-1-3">

            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default App;
