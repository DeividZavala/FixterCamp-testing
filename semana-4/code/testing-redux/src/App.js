import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Header from './components/Header';
import Card from './components/Card';
import Cart from './components/Cart';
import { getProducts } from './redux/productDuck';
import { getCart, editCart } from './redux/cartDuck';
import { denormalizeData } from './utils';
import './App.css';

function App() {
  const dispatch = useDispatch();
  const products = useSelector(state => denormalizeData(state.products.data));
  const status = useSelector(state => state.products.status);
  const error = useSelector(state => state.products.error);
  const cartItems = useSelector(state => state.cart.data);
  const items = denormalizeData(cartItems);

  useEffect(() => {
    if (status === '') dispatch(getProducts());
    if (status === '') dispatch(getCart());
  }, [dispatch]);

  const handleEdit = (item, type) => {
    if (item.id in cartItems) {
      if (type === 'delete' || (type === 'remove' && cartItems[item.id].quantity - 1 === 0)) {
        const filtered = Object.values(cartItems).filter(i => item.id !== i.id);
        const newOrder = { products: filtered };
        dispatch(editCart(newOrder));
      } else {
        const match = { ...cartItems[item.id] };
        const { quantity } = match;
        const newQuantity = type === 'add' ? quantity + 1 : quantity - 1;
        const edited = { ...match, quantity: newQuantity };
        const editedItems = { ...cartItems, [edited.id]: edited };
        const newOrder = { products: denormalizeData(editedItems) };
        dispatch(editCart(newOrder));
      }
    } else {
      const items = { ...cartItems, [item.id]: { ...item, quantity: 1 } };
      const newOrder = { products: denormalizeData(items) };
      dispatch(editCart(newOrder));
    }
  };

  return (
    <div className='App'>
      <Header items={items} />
      <div className='uk-section'>
        <div className='uk-container uk-container-expand'>
          <div className='uk-grid uk-grid-small uk-grid-match'>
            <div className='uk-width-2-3'>
              {status === 'fetching' && <div uk-spinner='ratio: 6'></div>}
              {status === 'finished' && (
                <div
                  id='products-container'
                  className='uk-grid uk-child-width-1-3 uk-grid-small uk-grid-match'
                >
                  {products.map((prod, index) => (
                    <Card key={index} {...prod} addItem={handleEdit} />
                  ))}
                </div>
              )}
              {status === 'error' && (
                <div
                  className='uk-alert-danger uk-flex uk-flex-middle uk-flex-center'
                  uk-alert='true'
                >
                  <p>{error}</p>
                </div>
              )}
            </div>
            <div className='uk-width-1-3'>
              <Cart items={items} editCart={handleEdit} />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default App;
