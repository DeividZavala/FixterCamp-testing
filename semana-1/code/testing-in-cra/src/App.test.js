import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";

describe("<App/>", () => {
  // fireEvent
  it("should add new items", () => {
    render(<App />);
    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "nuevo item" },
    });
    fireEvent.click(screen.getByRole("button"));

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "nuevo item 2" },
    });
    fireEvent.click(screen.getByRole("button"));
    expect(screen.getAllByRole("listitem").length).toBe(2);
  });
  it("should show error message if empty input", () => {
    render(<App />);
    fireEvent.click(screen.getByRole("button"));
    expect(screen.getByText("Debes agregar texto")).toBeInTheDocument();
  });

  // userEvent
  it("should add new items", () => {
    render(<App />);
    userEvent.type(screen.getByRole("textbox"), "nuevo item con user event");
    userEvent.click(screen.getByRole("button"));
    expect(screen.getAllByRole("listitem").length).toBe(1);
  });
  it("should show error message if empty input", () => {
    render(<App />);
    userEvent.click(screen.getByRole("button"));
    expect(screen.getByText("Debes agregar texto")).toBeInTheDocument();
  });
});
