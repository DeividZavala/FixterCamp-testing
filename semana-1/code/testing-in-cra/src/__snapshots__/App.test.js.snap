// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot <App /> should match snapshot if no data 1`] = `
<div
  className="App"
>
  <h1>
    Sin elementos para mostrar
  </h1>
  <div>
    <input
      placeholder="Ingresa el texto"
    />
    <button
      onClick={[Function]}
    >
      agregar
    </button>
  </div>
</div>
`;

exports[`<App /> snapshot <App /> should match snapshot when data 1`] = `
<div
  className="App"
>
  <ul>
    <li>
      test
    </li>
    <li>
      test 2
    </li>
  </ul>
  <div>
    <input
      placeholder="Ingresa el texto"
    />
    <button
      onClick={[Function]}
    >
      agregar
    </button>
  </div>
</div>
`;
